qemu/tests/migration_multi_host_with_reboot.py

```diff
@@ -0,0 +1,82 @@
+    for migrate method in class MultihostMigration.
+    steps:
+        1) boot vm
+        2) do reboot vm and keep reboot vm with system_reset,
+           let guest in bios stage, before kernel loading
+        3) do migration
+        4) login vm after migration
+
+    :param test: kvm test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+    preprocess_env = params.get("preprocess_env", "yes") == "yes"
+    mig_protocol = params.get("mig_protocol", "tcp")
+    mig_type = utils_test.qemu.MultihostMigration
+    if mig_protocol == "fd":
+        mig_type = utils_test.qemu.MultihostMigrationFd
+    if mig_protocol == "exec":
+        mig_type = utils_test.qemu.MultihostMigrationExec
+    if "rdma" in mig_protocol:
+        mig_type = utils_test.qemu.MultihostMigrationRdma
+
+    class TestMultihostMigration(mig_type):
+
+        def __init__(self, test, params, env):
+            super(TestMultihostMigration, self).__init__(test, params, env,
+                                                         preprocess_env)
+            self.srchost = self.params.get("hosts")[0]
+            self.dsthost = self.params.get("hosts")[1]
+            self.is_src = params["hostid"] == self.srchost
+            self.vms = params["vms"].split()
+            self.vm = params["vms"].split()[0]
+            self.login_timeout = int(params.get("login_timeout", 360))
+
+        def before_migration(self, mig_data):
+            def do_reboot(vm):
+                reboot_method = mig_data.params.get("reboot_method",
+                                                    "system_reset")
+                reboot_timeout = float(mig_data.params.get("reboot_timeout", 30))
+                if self.is_src:
+                    logging.info("Do '%s' before migraion..." % reboot_method)
+
+                    end_time = time.time() + reboot_timeout
+                    while time.time() < end_time:
+                        for vm in mig_data.vms:
+                            vm.monitor.clear_event("RESET")
+                            vm.monitor.cmd(reboot_method)
+                            reseted = utils_misc.wait_for(lambda:
+                                                          vm.monitor.get_event("RESET"),
+                                                          timeout=self.login_timeout)
+                            if not reseted:
+                                raise error.TestFail("Not found RESET event after "
+                                                     "execute 'system_reset'")
+                            vm.monitor.clear_event("RESET")
+                        time.sleep(2)
+
+            vm = env.get_vm(params["main_vm"])
+            bg = utils.InterruptedThread(do_reboot, (vm,))
+            bg.start()
+
+        def migration_scenario(self, worker=None):
+            if params.get("check_vm_before_migration", "yes") == "no":
```

Comment:
"start_migration_timeout" will be meaningless if run migration immediately when guest boot up

and need to add error.content for migration step